hera_cal/abscal.py

```diff
@@ -1291,6 +1295,280 @@ def global_phase_slope_logcal(model, data, antpos, reds=None, solver='linfit', w
+        Array of normalized data/model ratios. Shape (n_bls,).
+    weights : np.ndarray or jnp.ndarray
+        Array of weights. Shape (n_bls,).
+    
+    Returns
+    -------
+    Z : float
+        Metric of how well the model and data can be made to match after abscal.
+    """
+    phase = jnp.sum(x * b_vecs, axis=1)
+    phasor = jnp.cos(phase) + 1j * jnp.sin(phase)
+    Z = jnp.mean(phasor * normalized_data_model_ratio * weights)
+    return Z
+
+@jax.jit
+def _grad_and_hess_real(x, b_vec, w_expi_psi):
+    """
+    Evaluates the gradient and hessian for a given set of input parameters, x.
+
+    Parameters:
+    -----------
+    x : np.ndarray or jnp.ndarray
+        Solution vector of the tip-tilt phase slopes. Shape (n_dims,).
+    b_vec : np.ndarray or jnp.ndarray
+        Array of baseline vectors. Shape (n_bls, n_dims).
+    w_expi_psi : np.ndarray or jnp.ndarray
+        Array of weighted normalized data/model ratios. Shape (n_bls,).
+
+    Returns:
+    --------
+    grad : np.ndarray or jnp.ndarray
+        Gradient of the objective function. Shape (n_dims,).
+    hess : np.ndarray or jnp.ndarray
+        Hessian of the objective function. Shape (n_dims, n_dims).
+    """
+    x_dot_b = jnp.dot(b_vec, x)
+    cos_xb = jnp.cos(x_dot_b)
+    sin_xb = jnp.sin(x_dot_b)
+    w_cos_d = w_expi_psi.real
+    w_sin_d = w_expi_psi.imag
+    w_sin_d_xb = cos_xb * w_sin_d + sin_xb * w_cos_d
+    w_cos_d_xb = cos_xb * w_cos_d - sin_xb * w_sin_d
+    grad = -jnp.mean(w_sin_d_xb[..., None] * b_vec, axis=0)
+    hess = -jnp.mean(b_vec[:, None, :] * b_vec[..., None] * w_cos_d_xb[..., None, None], axis=0)
+    return grad, hess
+
+@jax.jit
+def _newton_solve_real(x0, transformed_b_vecs, normalized_data_model_ratio, weights, tol=1e-8, maxiter=25):
+    """
+    Refines a delay-slope solution (x0) to calibrate the phase slopes seen in normalized_data_model_ratio.
+
+    Parameters:
+    -----------
+    x0: np.ndarray or jnp.ndarray
+        Initial guess for the phase slopes to be calibrated.
+    transformed_b_vecs: np.ndarray or jnp.ndarray
+        Array of baseline vectors in the transformed coordinate system.
+    normalized_data_model_ratio: np.ndarray or jnp.ndarray
+        Array of normalized data-model ratios.
+    weights: np.ndarray or jnp.ndarray
+        Array of weights for each baseline.
+    tol: float
+        Tolerance for the convergence of the Newton solver.
+    maxiter: int
+        Maximum number of iterations for the Newton solver.
+
+    Returns:
+    --------
+    x: jnp.ndarray
+        Array of refined phase slopes.
+    i: int
+        Number of iterations performed by the Newton solver.
+    """
+    def inner_function(args):
+        """
+        Function evaluated on each iteration of the while loop
+        """
+        x, i, step_diff = args
+        grad, hess = _grad_and_hess_real(x, transformed_b_vecs, w_ndmr)
+        step = -1 * jnp.linalg.solve(hess, grad)
```

Comment:
Question for @zacharymartinot: Is there any reason to step less than -1 * the hessian (i.e. a damped iterator?)

Reply:
I did try this in another implementation of this function. I could just make the change and add it as a keyword argument.

Reply:
I thought about this briefly, but not very carefully. Its possible that including a damping factor could be useful when we're at the hairy edge of the model being "good enough". But I'm pretty sure a constant damping factor isn't going to do anything than add iterations in general, it would have to be something computed per step. and figuring that out would be a small project, if there's anything there at all. And if you're going down that road you might as well consider alternatives to newton's method entirely.

So until someone actually investigates, keep it simple. No damping.

Reply:
Yeah, sounds like an unnecessary rabbit hole... just something I thought I'd mention.

Reply:
I thought about this briefly, but not very carefully. Its possible that including a damping factor could be useful when we're at the hairy edge of the model being "good enough". But I'm pretty sure a constant damping factor isn't going to do anything than add iterations in general, it would have to be something computed per step. and figuring that out would be a small project, if there's anything there at all. And if you're going down that road you might as well consider alternatives to newton's method entirely.

So until someone actually investigates, keep it simple. No damping.